drivers/mipi_dbi/mipi_dbi_gpio.c

```diff
@@ -0,0 +1,373 @@
+	if (ret < 0) {
+		return ret;
+	}
+	k_sleep(delay);
+	return gpio_pin_set_dt(&config->reset, 0);
+}
+
+static int mipi_dbi_gpio_release(const struct device *dev, const struct mipi_dbi_config *dbi_config)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(dbi_config);
+
+	return 0;
+}
+
+static int mipi_dbi_gpio_init(const struct device *dev)
+{
+	const struct mipi_dbi_gpio_config *config = dev->config;
+	const char *failed_pin = NULL;
+	int ret = 0;
+#if MIPI_DBI_8_BIT_MODE
+	struct mipi_dbi_gpio_data *data = dev->data;
+#endif
+
+	if (gpio_is_ready_dt(&config->cmd_data)) {
+		ret = gpio_pin_configure_dt(&config->cmd_data, GPIO_OUTPUT_ACTIVE);
+		if (ret < 0) {
+			failed_pin = "cmd_data";
+			goto fail;
+		}
+		gpio_pin_set_dt(&config->cmd_data, 0);
+	}
+	if (gpio_is_ready_dt(&config->rd)) {
+		gpio_pin_configure_dt(&config->rd, GPIO_OUTPUT_ACTIVE);
+		/* Don't emit an error because this pin is unused in type A */
+		gpio_pin_set_dt(&config->rd, 1);
+	}
+	if (gpio_is_ready_dt(&config->wr)) {
+		ret = gpio_pin_configure_dt(&config->wr, GPIO_OUTPUT_ACTIVE);
+		if (ret < 0) {
+			failed_pin = "wr";
+			goto fail;
+		}
+		gpio_pin_set_dt(&config->wr, 1);
+	}
+	if (gpio_is_ready_dt(&config->e)) {
+		gpio_pin_configure_dt(&config->e, GPIO_OUTPUT_ACTIVE);
+		/* Don't emit an error because this pin is unused in type B */
+		gpio_pin_set_dt(&config->e, 0);
+	}
+	if (gpio_is_ready_dt(&config->cs)) {
+		ret = gpio_pin_configure_dt(&config->cs, GPIO_OUTPUT_ACTIVE);
+		if (ret < 0) {
+			failed_pin = "cs";
+			goto fail;
+		}
+		gpio_pin_set_dt(&config->cs, 0);
+	}
+	if (gpio_is_ready_dt(&config->reset)) {
+		ret = gpio_pin_configure_dt(&config->reset, GPIO_OUTPUT_ACTIVE);
+		if (ret < 0) {
+			failed_pin = "reset";
+			goto fail;
+		}
+		gpio_pin_set_dt(&config->reset, 0);
+	}
+	for (int i = 0; i < config->data_bus_width; i++) {
+		if (gpio_is_ready_dt(&config->data[i])) {
+			ret = gpio_pin_configure_dt(&config->data[i], GPIO_OUTPUT_ACTIVE);
+			if (ret < 0) {
+				failed_pin = "data";
+				goto fail;
+			}
+			gpio_pin_set_dt(&config->data[i], 0);
+		}
+	}
+
+#if MIPI_DBI_8_BIT_MODE
+	/* To optimize performance, we test whether all the data pins are
+	 * on the same port. If they are, we can set the whole port in one go
+	 * instead of setting each pin individually.
+	 * For 8-bit mode only because LUT size grows exponentially.
+	 */
+	if (config->data_bus_width == 8) {
+		data->port = config->data[0].port;
+		data->single_port = true;
+		data->data_mask = 0;
+		for (int i = 0; i < config->data_bus_width; i++) {
+			if (data->port != config->data[i].port) {
+				data->single_port = false;
+			}
+			data->data_mask |= (1 << config->data[i].pin);
+		}
+
+		if (data->single_port) {
+			LOG_DBG("LUT optimization enabled. data_mask=0x%x", data->data_mask);
+
+			/* Calculate a look-up table to support arbitrary bit orders. */
+			for (int i = 0; i < (1 << config->data_bus_width); i++) {
+				data->data_lut[i] = 0;
+				for (int j = 0; j < config->data_bus_width; j++) {
```

Comment:
I *think* we could generate this LUT statically like so:

```
#define LUT_GEN(i, _) ((i & (1 << 0)) ? DT_INST_GPIO_PIN_BY_IDX(n, data_gpios, 0) : 0) \
                        ((i & (1 << 1)) ? DT_INST_GPIO_PIN_BY_IDX(n, data_gpios, 1) : 0) \
                        ((i & (1 << 2)) ? DT_INST_GPIO_PIN_BY_IDX(n, data_gpios, 2) : 0)
                        /* ... continue up to 8 wide data bus ... */

uint32_t lut[] = {
        LISTIFY(256, LUT_GEN, (,))
};
```

Not proud of these macrobatics, but I think it might allow us to save the 1Kib of RAM here.

Reply:
Done.